tao_compiler/mlir/disc/transforms/disc_gpu_source_to_lib.cc

```diff
@@ -0,0 +1,266 @@
+      source_code_ops.push_back(source_code_op);
+    });
+
+    for (auto source_code_op : source_code_ops) {
+      auto code = source_code_op.getCode().str();
+      if (code.empty()) {
+        signalPassFailure();
+        return;
+      }
+      cuda_code += code;
+      // if (auto attr = source_code_op.getCode()) {
+      // cuda_code += attr.getValue();
+      // } else {
+      // signalPassFailure();
+      // return;
+      // }
+    }
+
+    llvm::errs() << "[ZZ] cuda code to compile:\n" << cuda_code << "\n";
+
+    // Compile CUDA code.
+    std::string bin_path;
+    if (failed(compilePreprocessedCUDASourceToLib(cuda_code, bin_path))) {
+      signalPassFailure();
+      return;
+    }
+
+    for (auto source_code_op : source_code_ops) {
+      OpBuilder builder(source_code_op);
+      source_code_op->setAttr(kDynLibPathAttr, builder.getStringAttr(bin_path));
+    }
+  }
+
+ private:
+  int cc_major_;
+  int cc_minor_;
+
+ private:
+  // Return true on success, false otherwise.
+  bool executeCommand(const std::string& cmd, std::string* output);
+  LogicalResult compilePreprocessedCUDASourceToLib(const std::string& source,
+                                                   std::string& bin_path);
+};
+
+bool DiscGPUSourceToLibPass::executeCommand(const std::string& cmd,
+                                            std::string* output) {
+  FILE* pipe = popen(cmd.c_str(), "r");
+  if (!pipe) {
+    return false;
+  }
+
+  if (output) {
+    char tmp_buffer[128];
+    while (fgets(tmp_buffer, sizeof(tmp_buffer), pipe)) {
+      output->append(tmp_buffer);
+    }
+  }
+  return pclose(pipe) == 0;
+}
+
+LogicalResult DiscGPUSourceToLibPass::compilePreprocessedCUDASourceToLib(
+    const std::string& source, std::string& bin_path) {
+  std::string random_number = std::to_string(tensorflow::random::New64());
+  std::string tmp_path = "/tmp/";
+  bin_path = tmp_path + random_number + ".so";
+
+  // clang-format off
+  const char *headers_cpp1 =
+#include "tensorflow/compiler/mlir/disc/utils/cutlass_header_preprocess.cpp1.ii.h"
+  ;
+  const char *headers_cpp4 =
+#include "tensorflow/compiler/mlir/disc/utils/cutlass_header_preprocess.cpp4.ii.h"
+  ;
+  // clang-format on
+
+  std::ofstream cufile_cpp1;
+  std::string source_path_cpp1 = tmp_path + random_number + ".cpp1.ii";
+  cufile_cpp1.open(source_path_cpp1);
+  cufile_cpp1 << headers_cpp1;
+  cufile_cpp1 << source;
+  cufile_cpp1.close();
+
+  std::ofstream cufile_cpp4;
+  std::string source_path_cpp4 = tmp_path + random_number + ".cpp4.ii";
+  cufile_cpp4.open(source_path_cpp4);
+  cufile_cpp4 << headers_cpp4;
+  cufile_cpp4 << source;
+  cufile_cpp4.close();
+
+  std::vector<std::string> compile_trajectory = compile_trajectory_template;
+  std::string cuda_arch = std::to_string(cc_major_ * 100 + cc_minor_ * 10);
+  std::string sm = std::to_string(cc_major_ * 10 + cc_minor_);
+  std::string cuda_home;
+  if (!executeCommand("dirname $(which nvcc)", &cuda_home) ||
```

Comment:
shall we share the same logic to find cuda toolkit?
https://github.com/pai-disc/tensorflow/blob/173d33c7c7aa8e3e9b352af2433f76ab8da2708e/tensorflow/tsl/platform/default/cuda_libdevice_path.cc#L31